lib/puppet/provider/nameservice.rb

```diff
@@ -276,6 +283,13 @@ def initialize(resource)
     super
     @custom_environment = {}
     @objectinfo = nil
+    if resource.is_a?(Hash) && !resource[:canonical_name].nil?
+      # we're instantiating from a hash, ie self.instances
+      @canonical_name = resource[:canonical_name]
+    else
+      # we're instantiating from a resource
+      @canonical_name = resource[:name]
```

Comment:
In this case, we've created a new resource object directly and might be testing that it exists, for example the puppet service user. in this case, we haven't defined a canonical name and its ok to just use name because we're not gathering this info from Etc and haven't done any scrubbing of invalid bytes